lib/IDE/CodeCompletionResultType.cpp

```diff
@@ -11,50 +11,233 @@
+      }
+      // Archetypes are also be used to model generic return types, in which
+      // case they don't have any conformsTo entries. We simply ignore those.
+      if (!USRConforms.empty()) {
+        ContextualTypes.emplace_back(USRConforms, /*IsConvertible=*/false);
+      }
+    }
+  }
+}
+
+TypeRelation
+USRBasedTypeContext::typeRelation(const USRBasedType *ResultType) const {
+  const USRBasedType *VoidType = Arena.getVoidType();
+  if (ResultType == VoidType) {
+    // Void is not convertible to anything and we don't report Void <-> Void
+    // identical matches (see USRBasedType::typeRelation). So we don't have to
+    // check anything if the result returns Void.
+    return TypeRelation::Unknown;
+  }
+
+  TypeRelation Res = TypeRelation::Unknown;
+  for (auto &ContextualType : ContextualTypes) {
+    Res = std::max(Res, ContextualType.typeRelation(ResultType, VoidType));
+  }
+  return Res;
+}
+
+// MARK: - USRBasedTypeArena
+
+USRBasedTypeArena::USRBasedTypeArena() {
+  // '$sytD' is the USR of the Void type.
+  VoidType = USRBasedType::fromUSR("$sytD", {}, *this);
+}
+
+const USRBasedType *USRBasedTypeArena::getVoidType() const { return VoidType; }
+
+// MARK: - USRBasedType
+
+const USRBasedType *USRBasedType::null(USRBasedTypeArena &Arena) {
+  return USRBasedType::fromUSR(/*USR=*/"", /*Supertypes=*/{}, Arena);
+}
+
+const USRBasedType *
+USRBasedType::fromUSR(StringRef USR, ArrayRef<const USRBasedType *> Supertypes,
+                      USRBasedTypeArena &Arena) {
+  auto ExistingTypeIt = Arena.CanonicalTypes.find(USR);
+  if (ExistingTypeIt != Arena.CanonicalTypes.end()) {
+    assert(ArrayRef<const USRBasedType *>(ExistingTypeIt->second->Supertypes) ==
+               Supertypes &&
+           "Same USR but different supertypes?");
+    return ExistingTypeIt->second;
+  }
+  // USR and Supertypes need to be allocated in the arena to be passed into the
+  // USRBasedType constructor. The elements of Supertypes are already allocated
+  // in the arena.
+  USR = USR.copy(Arena.Allocator);
+  Supertypes = Supertypes.copy(Arena.Allocator);
+
+  const USRBasedType *Result =
+      new (Arena.Allocator) USRBasedType(USR, Supertypes);
+  Arena.CanonicalTypes[USR] = Result;
+  return Result;
+}
+
+const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
+  // FIXME: We can't mangle archetypes, so we treat them as null USRBasedTypes.
+  if (!Ty || Ty->hasArchetype()) {
+    return USRBasedType::null(Arena);
+  }
+  // USRBasedTypes are backed by canonical types so that equivalent types have
+  // the same USR.
+  Ty = Ty->getCanonicalType();
+
+  SmallVector<const USRBasedType *, 2> Supertypes;
+  if (auto Nominal = Ty->getAnyNominal()) {
+    // Sorted conformances so we get a deterministic supertype order and can
+    // assert that USRBasedTypes with the same USR have the same supertypes.
+    auto Conformances = Nominal->getAllConformances(/*sorted=*/true);
```

Comment:
```
protocol P1 {}
protocol P2: P1 {}

class B: P2 {}

protocol Q1 {}
protocol Q2: Q1 {}

class C: B, Q2 {}
```
In this case, `USRBasedType::forType(C)` would be like
```
p1 = {"P1"}
p2 = {"P2", &p1}
b  = {"B", &p1, &p2}
q1 = {"Q1"}
q2 = {"Q2", &q1}
c  = {"C", &p1, &p2, &q1, &q2, &b}
```
right? This looks a waste of the space. Could you limit `Supertypes` only contains the "direct" super types? So the structure be like:
```
p1 = {"P1"}
p2 = {"P2", &p1}
b  = {"B", &p2}
q1 = {"Q1"}
q2 = {"Q2", &q1}
c  = {"C", &q2, &b}
```
When calculating the type relation, we should be able to collect all super type USRs by recursively look into the graph. Of course we need to be careful not to visit the same node twice, but it shouldn't be hard.

Reply:
Good idea. It might slow down the computation of USRBasedTypes a little but since that should only happen once to build up the cache, I think it’s a good tradeoff https://github.com/apple/swift/pull/41000/commits/daf8fb056196686f65b42ee0fd4d0fa641c7cc0c